### ISSUE TITLE
RuntimeError: coroutine raised StopIteration

### ISSUE THREAD

**Issue by davidblaisonneau-orange** [maas/python-libmaas#91]
I got an error at login phase: **RuntimeError: coroutine raised StopIteration** and i cannot find why.

I use libmaas in a virtualenv using the example:
`pip install git+https://github.com/maas/python-libmaas.git`

The code is pretty simple, and target a remote MAAS through a ssh port forwarding tunnel
```
#!/usr/bin/env python3.5

from maas.client import login
client = login(
    "http://localhost:5555/MAAS/",
    username="xxxx", password="xxxx",
)
```

Here is the output:
```
(amc) user@myhost:~⟫ python3 maas_client.py 
Traceback (most recent call last):
  File "/home/user/amc/lib/python3.5/site-packages/maas/client/bones/helpers.py", line 261, in _obtain_token
    cookie for cookie in session.cookie_jar
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "maas_client.py", line 7, in <module>
    username="xxxx", password="xxxx",
  File "/home/user/amc/lib/python3.5/site-packages/maas/client/utils/async.py", line 49, in wrapper
    result = eventloop.run_until_complete(result)
  File "/usr/lib/python3.5/asyncio/base_events.py", line 457, in run_until_complete
    return future.result()
  File "/usr/lib/python3.5/asyncio/futures.py", line 292, in result
    raise self._exception
  File "/usr/lib/python3.5/asyncio/tasks.py", line 239, in _step
    result = coro.send(None)
  File "/home/user/amc/lib/python3.5/site-packages/maas/client/__init__.py", line 43, in login
    url, username=username, password=password, insecure=insecure)
  File "/home/user/amc/lib/python3.5/site-packages/maas/client/viscera/__init__.py", line 603, in login
    url=url, username=username, password=password, insecure=insecure)
  File "/home/user/amc/lib/python3.5/site-packages/maas/client/bones/__init__.py", line 75, in login
    url=url, username=username, password=password, insecure=insecure)
  File "/home/user/amc/lib/python3.5/site-packages/maas/client/bones/helpers.py", line 199, in login
    url.geturl(), username, password, insecure=insecure)
RuntimeError: coroutine raised StopIteration
```

Here is the lib versions:
```
(amc) user@myhost:~⟫ pip freeze
aiohttp==1.3.5
appdirs==1.4.3
argcomplete==1.8.2
async-timeout==1.2.0
beautifulsoup4==4.5.3
chardet==2.3.0
colorclass==2.2.0
multidict==2.1.4
oauthlib==2.0.1
packaging==16.8
pkg-resources==0.0.0
pyparsing==2.2.0
python-libmaas==0.4.1
pytz==2016.10
PyYAML==3.12
six==1.10.0
terminaltables==3.1.0
yarl==0.9.8
```

Thanks for your help

**Comment by allenap** [comment]
The `StopIteration` problem is a weird bug that we noticed only a few days ago. We had thought this was related to the use of an IP address instead of a hostname when connecting to MAAS, but you're using `localhost` so it seems that the bug is more nuanced than that.

There's a workaround. Obtain, by hand, an API token from the user preferences page in the UI (`http://$address:$port/MAAS/account/prefs/`) and put it into the following command:

```python
from maas.client import connect
client = connect(
    "http://localhost:5555/MAAS/",
    apikey="$apikey",
)
```

I am working actively to fix the `login` bug. It needs a server-side fix too, and that's what I'm working on first. What version of the MAAS server are you using?

Gavin.

**Comment by davidblaisonneau-orange** [comment]
Thanks Gavin,

this workaround is good for me, and i prefer to connect through a key. I did not find any doc about connection with a key, did I missed it ? If there is not, it can be a good thing to add it in order to keep coherency with the 1.9 API Auth example (I can add a few line in your doc if you want)

My maas version is: 2.1.3+bzr5573-0ubuntu1 (16.04.1)

David

**Comment by allenap** [comment]
Thanks David.

I'm soon going to land a branch that will break `login` until a corresponding change in the MAAS server is released. Several people have had problems with `login` (it's _is_ a bit of a hack) so it will be no bad thing to direct people to `connect` straight away, and bring back `login` once it's done properly. Thank you for your offer to update the docs, but I have to update them imminently to talk about this stuff anyway. Contributions in general are welcome though!

Can you point me to the _1.9 API Auth example_ you mention? `python-libmaas` is intended only for MAAS 2.0+ so I worry that you and others are finding the wrong docs — which might be close enough in content to be plausible — and maybe there's something I can do to help people find their way to the correct docs.

Gavin.

**Comment by davidblaisonneau-orange** [comment]
The 1.9 page i refer is this one: https://docs.ubuntu.com/maas/1.9/en/api_authentication
I used this page before for our old maas servers.

David

**Comment by allenap** [comment]
This was fixed by #93, albeit in such a way that `login()` will not for anyone until we release a new 2.2 beta.
